overlord/devicestate/handlers_gadget.go

```diff
@@ -125,14 +144,35 @@ func (m *DeviceManager) doUpdateGadgetAssets(t *state.Task, _ *tomb.Tomb) error
 		return nil
 	}
 
+	// add kernel directories
+	currentKernelInfo, err := snapstate.CurrentInfo(st, groundDeviceCtx.Model().Kernel())
+	// XXX: switch to the normal `if err != nil { return err }` pattern
+	// here once all tests are updated and have a kernel
+	if err == nil {
+		currentData.KernelRootDir = currentKernelInfo.MountDir()
+		updateData.KernelRootDir = currentKernelInfo.MountDir()
+	}
+	// if this is a gadget update triggered by an updated kernel we
+	// need to ensure "updateData.KernelRootDir" points to the new kernel
+	if snapsup.Type == snap.TypeKernel {
+		updateKernelInfo, err := snap.ReadInfo(snapsup.InstanceName(), snapsup.SideInfo)
+		if err != nil {
+			return fmt.Errorf("cannot read candidate kernel snap details: %v", err)
+		}
+		updateData.KernelRootDir = updateKernelInfo.MountDir()
+	}
+
 	snapRollbackDir, err := makeRollbackDir(fmt.Sprintf("%v_%v", snapsup.InstanceName(), snapsup.SideInfo.Revision))
 	if err != nil {
 		return fmt.Errorf("cannot prepare update rollback directory: %v", err)
 	}
 
 	var updatePolicy gadget.UpdatePolicyFunc = nil
 
-	if isRemodel {
+	// Even with a remodel a kernel refresh only updates the kernel assets
```

Comment:
thanks for the change, do we test this combination somewhere?

Reply:
I added a test now